File: backend/app/schemas/analytics.py
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field, ConfigDict


class FinancialSummaryResponse(BaseModel):
    starting_balance: Decimal
    current_balance: Decimal
    total_income: Decimal
    total_expenses: Decimal
    net_cash_flow: Decimal
    income_transaction_count: int
    expense_transaction_count: int
    currency: str = "INR"

    model_config = ConfigDict(from_attributes=True)


class MonthlyAnalyticsResponse(BaseModel):
    year: int
    month: int
    monthly_income: Decimal
    monthly_expenses: Decimal
    monthly_net_cash_flow: Decimal
    transaction_count: int
    previous_month_income: Decimal
    previous_month_expenses: Decimal
    previous_month_net_cash_flow: Decimal
    income_change_percentage: Optional[Decimal] = None
    expense_change_percentage: Optional[Decimal] = None
    net_cash_flow_change_percentage: Optional[Decimal] = None

    model_config = ConfigDict(from_attributes=True)


class CategorySpendingItem(BaseModel):
    category: str
    amount: Decimal
    percentage: Decimal
    transaction_count: int

    model_config = ConfigDict(from_attributes=True)


class CategorySpendingResponse(BaseModel):
    year: int
    month: int
    total_expenses: Decimal
    items: List[CategorySpendingItem]

    model_config = ConfigDict(from_attributes=True)


class IncomeCategoryItem(BaseModel):
    category: str
    amount: Decimal
    percentage: Decimal
    transaction_count: int

    model_config = ConfigDict(from_attributes=True)


class IncomeCategoryResponse(BaseModel):
    year: int
    month: int
    total_income: Decimal
    items: List[IncomeCategoryItem]

    model_config = ConfigDict(from_attributes=True)


class DailyTrendItem(BaseModel):
    date: str = Field(..., description="Date in YYYY-MM-DD format")
    income: Decimal
    expenses: Decimal
    net: Decimal

    model_config = ConfigDict(from_attributes=True)


class DailyTrendResponse(BaseModel):
    year: int
    month: int
    days: List[DailyTrendItem]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/analytics.ts
export interface FinancialSummaryResponse {
  starting_balance: string
  current_balance: string
  total_income: string
  total_expenses: string
  net_cash_flow: string
  income_transaction_count: number
  expense_transaction_count: number
  currency: string
}

export interface MonthlyAnalyticsResponse {
  year: number
  month: number
  monthly_income: string
  monthly_expenses: string
  monthly_net_cash_flow: string
  transaction_count: number
  previous_month_income: string
  previous_month_expenses: string
  previous_month_net_cash_flow: string
  income_change_percentage: string | null
  expense_change_percentage: string | null
  net_cash_flow_change_percentage: string | null
}

export interface CategorySpendingItem {
  category: string
  amount: string
  percentage: string
  transaction_count: number
}

export interface CategorySpendingResponse {
  year: number
  month: number
  total_expenses: string
  items: CategorySpendingItem[]
}

export interface IncomeCategoryItem {
  category: string
  amount: string
  percentage: string
  transaction_count: number
}

export interface IncomeCategoryResponse {
  year: number
  month: number
  total_income: string
  items: IncomeCategoryItem[]
}

export interface DailyTrendItem {
  date: string
  income: string
  expenses: string
  net: string
}

export interface DailyTrendResponse {
  year: number
  month: number
  days: DailyTrendItem[]
}

export interface MonthFilterParams {
  year?: number
  month?: number
}
